File: backend/app/models/validation.py
"""Pydantic models for validation results."""

from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class ValidationCategory(str, Enum):
    """Category of validation check."""

    STRUCTURAL = "structural"  # Configuration structure issues
    DATA = "data"  # Issues found in actual data
    PERFORMANCE = "performance"  # Performance-related warnings


class ValidationPriority(str, Enum):
    """Priority level for validation checks."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ValidationError(BaseModel):
    """A validation error or warning."""

    severity: Literal["error", "warning", "info"] = Field(..., description="Severity level")
    entity: str | None = Field(default=None, description="Entity name where error occurred")
    field: str | None = Field(default=None, description="Field name where error occurred")
    message: str = Field(..., description="Error message")
    code: str | None = Field(default=None, description="Error code for programmatic handling")
    suggestion: str | None = Field(default=None, description="Suggested fix")
    category: ValidationCategory = Field(default=ValidationCategory.STRUCTURAL, description="Validation category")
    priority: ValidationPriority = Field(default=ValidationPriority.MEDIUM, description="Validation priority")
    auto_fixable: bool = Field(default=False, description="Whether this issue can be auto-fixed")


class ValidationResult(BaseModel):
    """Result of configuration validation."""

    is_valid: bool = Field(..., description="Whether configuration is valid")
    errors: list[ValidationError] = Field(default_factory=list, description="Validation errors")
    warnings: list[ValidationError] = Field(default_factory=list, description="Validation warnings")
    info: list[ValidationError] = Field(default_factory=list, description="Informational messages")
    error_count: int = Field(default=0, description="Number of errors")
    warning_count: int = Field(default=0, description="Number of warnings")

    @model_validator(mode="after")
    def calculate_counts(self) -> "ValidationResult":
        """Calculate error and warning counts from lists."""
        self.error_count = len(self.errors)
        self.warning_count = len(self.warnings)
        return self

    @property
    def total_issues(self) -> int:
        """Total number of issues (errors + warnings + info)."""
        return self.error_count + self.warning_count + len(self.info)

    def get_errors_for_entity(self, entity_name: str) -> list[ValidationError]:
        """Get all errors for a specific entity."""
        return [e for e in self.errors if e.entity == entity_name]

    def get_warnings_for_entity(self, entity_name: str) -> list[ValidationError]:
        """Get all warnings for a specific entity."""
        return [e for e in self.warnings if e.entity == entity_name]

File: frontend/src/types/validation.ts
/**
 * TypeScript types for validation.
 */

export type ValidationSeverity = 'error' | 'warning' | 'info'
export type ValidationCategory = 'structural' | 'data' | 'performance' | 'conformance'
export type ValidationPriority = 'low' | 'medium' | 'high' | 'critical'
export type DataValidationMode = 'sample' | 'complete'

export interface ValidationError {
  severity: ValidationSeverity
  entity?: string | null
  branch_name?: string | null
  branch_source?: string | null
  field?: string | null
  message: string
  code?: string | null
  suggestion?: string | null
  category?: ValidationCategory
  priority?: ValidationPriority
  auto_fixable?: boolean
}

export interface ValidationBranchGroup {
  key: string
  branchName: string | null
  branchSource: string | null
  messages: ValidationError[]
}

export interface ValidationEntityGroup {
  key: string
  entity: string | null
  messages: ValidationError[]
  entityMessages: ValidationError[]
  branchGroups: ValidationBranchGroup[]
}

export interface ValidationResult {
  is_valid: boolean
  errors: ValidationError[]
  warnings: ValidationError[]
  info: ValidationError[]
  error_count: number
  warning_count: number
  validation_mode?: DataValidationMode | null
}

// Helper functions
export function getErrorCount(result: ValidationResult): number {
  return result.errors.length
}

export function getWarningCount(result: ValidationResult): number {
  return result.warnings.length
}

export function getTotalIssues(result: ValidationResult): number {
  return result.errors.length + result.warnings.length
}

export function getErrorsForEntity(result: ValidationResult, entityName: string): ValidationError[] {
  return result.errors.filter((e) => e.entity === entityName)
}

export function getWarningsForEntity(result: ValidationResult, entityName: string): ValidationError[] {
  return result.warnings.filter((e) => e.entity === entityName)
}

export function groupByEntity(errors: ValidationError[]): Map<string, ValidationError[]> {
  const grouped = new Map<string, ValidationError[]>()
  for (const error of errors) {
    const key = error.entity || 'general'
    if (!grouped.has(key)) {
      grouped.set(key, [])
    }
    grouped.get(key)!.push(error)
  }
  return grouped
}

export function groupByEntityScope(messages: ValidationError[]): ValidationEntityGroup[] {
  const entityGroups = new Map<string, ValidationEntityGroup>()

  for (const message of messages) {
    const entityKey = message.entity || 'general'
    if (!entityGroups.has(entityKey)) {
      entityGroups.set(entityKey, {
        key: entityKey,
        entity: message.entity || null,
        messages: [],
        entityMessages: [],
        branchGroups: [],
      })
    }

    const entityGroup = entityGroups.get(entityKey)!
    entityGroup.messages.push(message)

    if (!message.branch_name && !message.branch_source) {
      entityGroup.entityMessages.push(message)
      continue
    }

    const branchKey = `${message.branch_name || ''}::${message.branch_source || ''}`
    let branchGroup = entityGroup.branchGroups.find((group) => group.key === branchKey)
    if (!branchGroup) {
      branchGroup = {
        key: branchKey,
        branchName: message.branch_name || null,
        branchSource: message.branch_source || null,
        messages: [],
      }
      entityGroup.branchGroups.push(branchGroup)
    }

    branchGroup.messages.push(message)
  }

  return Array.from(entityGroups.values()).sort((left, right) => {
    if (left.entity === null) return -1
    if (right.entity === null) return 1
    return left.key.localeCompare(right.key)
  })
}

export function groupBySeverity(result: ValidationResult): Record<ValidationSeverity, ValidationError[]> {
  return {
    error: result.errors,
    warning: result.warnings,
    info: result.info,
  }
}
